File: trading-engine/src/trading_engine/risk_engine.py
"""リスクチェックエンジン

全注文は必ずここを通過する。1つでも違反があれば注文を拒否。
"""

from dataclasses import dataclass
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

import sys

sys.path.insert(0, str(__import__("pathlib").Path(__file__).resolve().parents[3] / "mcp-server" / "src"))
from claude_trade.models import RiskState, Order  # noqa: E402

from trading_engine.config import settings
from trading_engine.portfolio_calc import OrderRequest

engine = create_async_engine(settings.database_url, echo=False)
async_session = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)


@dataclass
class RiskCheckResult:
    approved: bool
    checks: dict[str, bool]
    reasons: list[str]


async def get_risk_state() -> RiskState:
    async with async_session() as session:
        result = await session.execute(select(RiskState).where(RiskState.id == 1))
        state = result.scalar_one_or_none()
        if state is None:
            state = RiskState(id=1)
            session.add(state)
            await session.commit()
            await session.refresh(state)
        return state


async def check_order(order: OrderRequest, nav: float, current_price: float) -> RiskCheckResult:
    """個別注文のリスクチェック"""
    checks = {}
    reasons = []

    risk_state = await get_risk_state()

    # 1. Kill switch
    checks["kill_switch"] = not risk_state.kill_switch_active
    if risk_state.kill_switch_active:
        reasons.append(f"Kill switch active: {risk_state.kill_switch_reason}")

    # 2. Live trading disabled check
    checks["trading_enabled"] = not risk_state.live_trading_enabled or True  # paper always ok
    # Note: actual live trading guard is in executor

    # 3. Max position size
    order_value = order.quantity * current_price
    max_value = nav * (settings.max_position_pct / 100.0)
    checks["position_size"] = order_value <= max_value
    if not checks["position_size"]:
        reasons.append(
            f"Order value ${order_value:.0f} exceeds max ${max_value:.0f} "
            f"({settings.max_position_pct}% of NAV)"
        )

    # 4. Daily order count
    today = datetime.utcnow().strftime("%Y-%m-%d")
    if risk_state.last_reset_date != today:
        risk_state.daily_order_count = 0
        risk_state.last_reset_date = today

    checks["daily_orders"] = risk_state.daily_order_count < settings.max_daily_orders
    if not checks["daily_orders"]:
        reasons.append(f"Daily order limit reached: {risk_state.daily_order_count}/{settings.max_daily_orders}")

    # 5. Daily loss limit
    checks["daily_loss"] = risk_state.daily_loss_pct < settings.daily_loss_limit_pct
    if not checks["daily_loss"]:
        reasons.append(f"Daily loss {risk_state.daily_loss_pct:.2f}% exceeds limit {settings.daily_loss_limit_pct}%")

    approved = all(checks.values())
    return RiskCheckResult(approved=approved, checks=checks, reasons=reasons)


async def increment_order_count():
    """日次注文カウントをインクリメント"""
    async with async_session() as session:
        result = await session.execute(select(RiskState).where(RiskState.id == 1))
        state = result.scalar_one_or_none()
        if state:
            state.daily_order_count += 1
            state.updated_at = datetime.utcnow()
            await session.commit()

File: packages/server/src/trading-engine/risk-engine.ts
/** リスクチェックエンジン — 全注文は必ずここを通過する */

import { eq } from "drizzle-orm";
import type { OrderRequest } from "../broker/types.js";
import { loadEnv } from "../config.js";
import { db } from "../db/client.js";
import { type RiskState, riskState } from "../db/schema.js";

export interface RiskCheckResult {
  approved: boolean;
  checks: Record<string, boolean>;
  reasons: string[];
}

export async function getRiskState(): Promise<RiskState> {
  const [state] = await db.select().from(riskState).where(eq(riskState.id, 1));

  if (!state) {
    const [created] = await db.insert(riskState).values({ id: 1 }).returning();
    return created;
  }
  return state;
}

export async function checkOrder(
  order: OrderRequest,
  nav: number,
  currentPrice: number,
): Promise<RiskCheckResult> {
  const env = loadEnv();
  const checks: Record<string, boolean> = {};
  const reasons: string[] = [];
  const state = await getRiskState();

  // 1. Kill switch
  checks.kill_switch = !state.killSwitchActive;
  if (state.killSwitchActive) {
    reasons.push(`Kill switch active: ${state.killSwitchReason}`);
  }

  // 2. Live trading disabled check
  checks.trading_enabled = true; // paper always ok

  // 3. Max position size
  const orderValue = order.quantity * currentPrice;
  const maxValue = nav * (env.MAX_POSITION_PCT / 100);
  checks.position_size = orderValue <= maxValue;
  if (!checks.position_size) {
    reasons.push(
      `Order value ${orderValue.toFixed(0)} exceeds max ${maxValue.toFixed(0)} (${env.MAX_POSITION_PCT}% of NAV)`,
    );
  }

  // 4. Daily order count
  const today = new Date().toISOString().slice(0, 10);
  let dailyCount = state.dailyOrderCount ?? 0;
  if (state.lastResetDate !== today) {
    dailyCount = 0;
  }
  checks.daily_orders = dailyCount < env.MAX_DAILY_ORDERS;
  if (!checks.daily_orders) {
    reasons.push(`Daily order limit reached: ${dailyCount}/${env.MAX_DAILY_ORDERS}`);
  }

  // 5. Daily loss limit
  checks.daily_loss = (state.dailyLossPct ?? 0) < env.DAILY_LOSS_LIMIT_PCT;
  if (!checks.daily_loss) {
    reasons.push(
      `Daily loss ${(state.dailyLossPct ?? 0).toFixed(2)}% exceeds limit ${env.DAILY_LOSS_LIMIT_PCT}%`,
    );
  }

  const approved = Object.values(checks).every(Boolean);
  return { approved, checks, reasons };
}

export async function incrementOrderCount(): Promise<void> {
  const state = await getRiskState();
  const today = new Date().toISOString().slice(0, 10);
  const newCount = state.lastResetDate === today ? (state.dailyOrderCount ?? 0) + 1 : 1;

  await db
    .update(riskState)
    .set({
      dailyOrderCount: newCount,
      lastResetDate: today,
      updatedAt: new Date(),
    })
    .where(eq(riskState.id, 1));
}
